File: app/frontend/lib/logo-upload.ts
import policy from '../../../config/logo_upload.json';

export { policy as logoPolicy };
export type LogoCheck = { error: string | null; width?: number; height?: number };

export function logoDimensionsError(width: number, height: number): string | null {
  if (width * height > policy.maxPixels) return policy.errors.large;
  if (Math.max(width, height) < policy.minLongSide || Math.min(width, height) < policy.minShortSide) return policy.errors.small;
  return null;
}

export async function checkLogo(file: File): Promise<LogoCheck> {
  if (file.size > policy.maxBytes) return { error: policy.errors.size };
  if (!policy.types.includes(file.type)) return { error: policy.errors.type };
  const url = URL.createObjectURL(file);
  try {
    const image = new Image();
    image.src = url;
    await image.decode();
    const width = image.naturalWidth;
    const height = image.naturalHeight;
    return { error: logoDimensionsError(width, height), width, height };
  } catch {
    return { error: policy.errors.invalid };
  } finally {
    URL.revokeObjectURL(url);
  }
}

File: app/lib/logo_upload.rb
# Shared with the browser. Inspect the bytes before writing anything to Active Storage;
# dimensions are a resolution floor, not a promise that an upscaled image is sharp.
class LogoUpload
  POLICY = JSON.parse(Rails.root.join("config/logo_upload.json").read).freeze

  def self.error_for(file)
    return POLICY["errors"]["size"] if file.size > POLICY["maxBytes"]

    file.rewind
    bytes = file.read
    file.rewind
    type = Marcel::MimeType.for(StringIO.new(bytes))
    return POLICY["errors"]["type"] unless POLICY["types"].include?(type)

    image = Vips::Image.new_from_buffer(bytes, "", access: :sequential, fail_on: :warning)
    return POLICY["errors"]["large"] if image.width * image.height > POLICY["maxPixels"]
    return POLICY["errors"]["small"] if [image.width, image.height].max < POLICY["minLongSide"] || [image.width, image.height].min < POLICY["minShortSide"]

    # Vips is lazy: force decoding so a valid header with truncated pixels is rejected.
    image.avg
    nil
  rescue Vips::Error
    POLICY["errors"]["invalid"]
  end
end
